sdk/python/cmd/pulumi-language-python-exec

```diff
@@ -107,14 +108,38 @@ if __name__ == "__main__":
     logging.getLogger("asyncio").setLevel(logging.CRITICAL)
     exit_code = 1
     try:
+        # record the location of the user's program to return user tracebacks
+        user_program_abspath = os.path.abspath(os.path.join(args.PROGRAM, '__main__'))
+
         coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))
         loop.run_until_complete(coro)
         exit_code = 0
     except pulumi.RunError as e:
         pulumi.log.error(str(e))
     except Exception:
         # provide the user with traceback information for troubleshooting
-        error_msg = "Program failed with an unhandled exception:\n" + traceback.format_exc()
+        def get_user_stacktrace() -> str:
+            def get_abs_module_path(mod_path):
+                return os.path.abspath(os.path.splitext(mod_path)[0])
+
+            tb = traceback.extract_tb(sys.exc_info()[2])
+
+            for frame_index, frame in enumerate(tb):
+                # loop over stack frames until we reach the main program
+                # then return the traceback truncated to the user's code
+                cur_module = frame[0]
+
+                if get_abs_module_path(user_program_abspath) == get_abs_module_path(cur_module):
+                    # we have detected the start of a user's stack trace
```

Comment:
The attempt to detect the start of the user stack trace seems quite fragile. It it always __main__? Are there other ways we could run the detection?

Reply:
> If I specify in Pulumi.yaml
> 
> ```
> main: a.py
> ```
> 
> Then `__main__` is not the startup module anymore.

I'm unsure because from the `Pulumi.yaml` file reference:
> main: (optional) an override for the main program’s location. By default, the program’s working directory is assumed to be the location of Pulumi.yaml. To choose a different location, use the main property. For example, if your Pulumi program is in a subdirectory infra/, use main: infra/.

https://www.pulumi.com/docs/reference/pulumi-yaml/

> One idea that seems good to me is instead to look at a stack frame from `runpy.py` that does `run_path` because that's what we call right here, `pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))`; if we found that stack trace, we skip everything up to that stack trace plus any subsequent stack traces from `runpy`. This detection would not need to worry about `__main__` naming.

this binds our correctness to the internal implementation of `runpy` which I think is more fragile because it's controlled by python and is internal implementation and could change whereas we have control/knowledge of the function that we're trying to run.

> Could we also think of a test here of some sort? The improvement is very nice but the detector is something that can easily go out of whack over time as Python changes. A test (esp unit test) would be very helpful here.

A unit test would be good for this yes and could detect if the runpy implementation drifts if we go with that, but could also ensure that Pulumi.yaml and the program detection are in step.

Reply:
I'm pretty sure because I ran a test program. main: a.py renames entry-point. I also recall fixing bugs in this area. There's users doing this (renaming entrypoint).

Note that only your additions assume the entrypoint file is called __main__.py; in this code run_name parameter reconfigures the global var __name__ in the script, not the path to the script.

```
        coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__')
```

If you could please run this logic through a test program that species `main: a.py` and see if your changes still work with that, that'd be great.

> this binds our correctness to the internal implementation of runpy

True, but I bet runpy changes very slowly and would mostly call itself. The unit test would catch when this changes. I can concede to the current solution of scanning for __main__.py if it happens to work for programs that modify the entrypoint.

> A unit test would be good 

We'd want that regardless what we do here. 👍

Reply:
Ok I can fix the `main: a.py` details. I'll also update the hugo documentation to include this detail.

Reply:
added tests